### PR TITLE
Make the component in coefficients an optional type

### PR DESCRIPTION
**Pull request by jamiebramwell**
The removes the `component = -1` magic number in the coefficient definition and instead uses an optional type. This propagated through the `BoundaryCondition`'s and into the physics modules.

### CODE REVIEW

**Review by joshessman-llnl — APPROVED** [review]
👍 for nullable types!

File: src/serac/physics/nonlinear_solid.cpp [hunk 1]
```diff
@@ -93,8 +93,11 @@ NonlinearSolid::NonlinearSolid(std::shared_ptr<mfem::ParMesh> mesh, const Nonlin
         auto disp_coef = std::make_shared<mfem::VectorFunctionCoefficient>(bc.coef_opts.constructVector(dim));
         setDisplacementBCs(bc.attrs, disp_coef);
       } else {
+        SLIC_ERROR_ROOT_IF(!bc.coef_opts.component,
+                           "Component not specified with scalar coefficient when setting the displacement condition.",
+                           mpi_rank_);
```
Comment (white238) [inline src/serac/physics/nonlinear_solid.cpp]: rank goes second.

```
#define SLIC_ERROR_ROOT_IF(EXP, rank, msg)
```

Comment (jamiebramwell) [reply src/serac/physics/nonlinear_solid.cpp]: fixed. Thanks for the catch!

**Comment by jamiebramwell** [comment]
LGTM

**Review by white238 — APPROVED** [review]

**Comment by jamiebramwell** [comment]
LGTM

**Pull request merged**